### PR TITLE
Fix iterator bug for determinants with 32 orbitals or more

### PR DESCRIPTION
**Pull request by fevangelista**
This pull request replaces manual iterator arithmetic for the `BitArray` class (which had a bug when the number of orbitals was larger than 31). The old iterator is replaced with the safer `std::next`. New test cases have been added to increase coverage of edge cases.

**Iterator and permutation handling improvements:**

* Replaced manual pointer arithmetic with `std::next` in calls to `std::next_permutation` and related iterator assignments in `ci_spin_adapter.cc`, `ci_strings_makers.cc`, and `determinant_helpers.cc`, improving code clarity and reducing potential off-by-one errors.

**Code cleanup:**

* Removed a buggy implementation of `operator+` from the `BitArray::iterator` class in `bitarray.hpp`, simplifying the iterator implementation.

**Testing improvements:**

* Added new edge case tests to `test_determinant_helpers.py` for `hilbert_space` with larger input sizes, increasing test coverage and robustness.

### CODE REVIEW

File: tests/ci/test_determinant_helpers.py [hunk 1]
```diff
@@ -121,3 +121,9 @@ def test_determinant_hilbert_space_edge_cases():
     expected = [det("aa00"), det("a0a0"), det("a00a"), det("0aa0"), det("0a0a")]
     assert len(dets) == 5
     assert sorted(dets) == sorted(expected)
+
+    dets = hilbert_space(35, 1, 0)
+    assert len(dets) == 35
+
+    dets = hilbert_space(64, 0, 1)
```
[the remark this answers is not in the document]
Comment (fevangelista) [reply tests/ci/test_determinant_helpers.py:128]: The code changes affect the BitArray class iterator, which is not exposed to Python, so testing is a bit more tricky.
